src/units/unit.cpp

```diff
@@ -806,53 +806,114 @@ void unit::generate_traits(bool must_have_only)
 	LOG_UT << "Generating a trait for unit type " << type().log_id() << " with must_have_only " << must_have_only;
 	const unit_type& u_type = type();
 
-	// Calculate the unit's traits
 	config::const_child_itors current_traits = modifications_.child_range("trait");
-	std::vector<const config*> candidate_traits;
 
+	// Handle must-have only at the beginning
 	for(const config& t : u_type.possible_traits()) {
-		// Skip the trait if the unit already has it.
-		const std::string& tid = t["id"];
-		bool already = false;
-		for(const config& mod : current_traits) {
-			if(mod["id"] == tid) {
-				already = true;
-				break;
-			}
-		}
-
-		if(already) {
-			continue;
-		}
-
 		// Add the trait if it is mandatory.
 		const std::string& avl = t["availability"];
 		if(avl == "musthave") {
 			modifications_.add_child("trait", t);
 			current_traits = modifications_.child_range("trait");
 			continue;
 		}
-
-		// The trait is still available, mark it as a candidate for randomizing.
-		// For leaders, only traits with availability "any" are considered.
-		if(!must_have_only && (!can_recruit() || avl == "any")) {
-			candidate_traits.push_back(&t);
-		}
 	}
 
 	if(must_have_only) return;
 
+	std::vector<const config*> candidate_traits;
+	std::vector<std::string> temp_require_traits;
+	std::vector<std::string> temp_exclude_traits;
+
 	// Now randomly fill out to the number of traits required or until
 	// there aren't any more traits.
 	int nb_traits = current_traits.size();
 	int max_traits = u_type.num_traits();
-	for(; nb_traits < max_traits && !candidate_traits.empty(); ++nb_traits)
+	for(; nb_traits < max_traits; ++nb_traits)
 	{
+		current_traits = modifications_.child_range("trait");
+		candidate_traits.clear();
+		for(const config& t : u_type.possible_traits()) {
+			// Skip the trait if the unit already has it.
+			const std::string& tid = t["id"];
+			bool already = false;
+			for(const config& mod : current_traits) {
+				if(mod["id"] == tid) {
+					already = true;
+					break;
+				}
+			}
+
+			if(already)
+				continue;
```

Comment:
Doesn't look like we have an explicit rule about it but I would either write this on one line or use curly braces. The way it is now you might just add an indented statement and expect it to be behind the if...

Reply:
I will make sure to include braces, whenever I use if's. I personally find it rather appealing + less cluttering -> easier to read, so changing the code-style might be a nice idea